lib/storage/WebStorage.js

```diff
@@ -27,10 +32,15 @@ const webStorage = {
             .then(() => raiseStorageSyncEvent(key));
 
         // If we just call Storage.clear other tabs will have no idea which keys were available previously
-        // so that they can call keysChanged for them. That's why we iterate and remove keys one by one
-        this.clear = () => Storage.getAllKeys()
-            .then(keys => _.map(keys, key => this.removeItem(key)))
-            .then(tasks => Promise.all(tasks));
+        // so that they can call keysChanged for them. That's why we iterate over every key that doesn't need to be
+        // preserved and raise a storage sync event for them
+        this.clear = keyValuesToPreserve => Storage.getAllKeys()
+            .then((allKeys) => {
+                const keysToPreserve = _.map(keyValuesToPreserve, keyValueToPreserve => keyValueToPreserve[0]);
+                return _.filter(allKeys, key => !_.contains(keysToPreserve, key));
+            })
+            .then(keysToRemove => _.map(keysToRemove, key => raiseStorageSyncEvent(key)))
```

Comment:
If these are all the keys to be removed, why do we `raiseStorageSyncEvent` for each, before they are even cleared?

Why do we use `_.map` to iterate a list? 
`raiseStorageSyncEvent` does not return a promise, but even if it did, you'd still need to have something like

```js
.then(keysToRemove => {
  const promises = _.map(keysToRemove, key => raiseStorageSyncEvent(key));
  return Promise.all(promises);
})
```

I think it would be better to create a `CLEAR_ONYX` event, similarly to how `SYNC_ONYX` works

Instead of having `this.clear` work with `keyValuesToPreserve` it can work with just the `keysToPreserve`
This would allow us to raise a `CLEAR_ONYX` event with `keysToPreserve` as value 
Then before clearing storage
1. storage would multi get `keysToPreserve` 
2. then call storage clear
3. and finally multi set the result of `multiget` at point 1

This should be simpler and more efficient that iterating all the keys

Reply:
> Why do we use _.map to iterate a list?

You're right, it's not necessary, it's just what we were using before and it only needs to be a loop at a bare minimum.

> raiseStorageSyncEvent does not return a promise, but even if it did, you'd still need to have something like

Yes, that's almost exactly what was there before, as you can see in the diff, and it doesn't make any sense considering that `raiseStorageSyncEvent` doesn't return anything.

I'm not totally following the idea of `CLEAR_ONYX`, but the rest of what you said gave me some thoughts of a different way I can code this to keep this logic from having to deal with `keysToPreserve` at all. Let me think on it and I'll send in an update.

Reply:
> why do we raiseStorageSyncEvent for each, before they are even cleared?

Oh, to answer this... I am not sure. It's what was there before so I was trying to preserve the previous functionality (which I think was all added by yourself 😁 ).

I think it's so that other browser tabs will know that all the keys have been removed. Technically, it could happen either before or after calling `clear()`. If it's removed entirely, then other open tabs wouldn't sign out and they would retain all their existing data.

Reply:
> I think it's so that other browser tabs will know that all the keys have been removed. Technically, it could happen either before or after calling `clear()`

Yeah I was just confused about that part
I thought that because we call it before we clear storage, other tabs might have a chance to re-sync keys from storage and read values before we delete them 

So I think it would be safer to first call `Storage.clear()` and only then raise the storage sync events 

> I'm not totally following the idea of `CLEAR_ONYX`

What you've got now seems to be less changes overall, so if it there are no issues it doesn't matter